### PR TITLE
protobuf: add cmake builder, protobuf3_2{0,1}: init

### PR DESCRIPTION
**Pull request by jonringer**
###### Description of changes

Add cmake protobuf builder, so that we can expose `*.cmake` and `*.pc` files.

This is important because packages such as grpc expose .cmake files, which assume that protobuf exposes cmake files as well.

3.20 changelog: https://github.com/protocolbuffers/protobuf/releases/tag/v3.20.0

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
related upstream PR: https://github.com/protocolbuffers/protobuf/pull/9822

**Comment by bjornfor** [comment]
First commit message needs s/cmake/protobuf/: "cmake: add cmake generic builder"

**Comment by jonringer** [comment]
thanks @bjornfor, addressed :)

**Comment by jonringer** [comment]
added protobuf 3.21 as well

**Comment by jonringer** [comment]
I'm going to merge this in a few days. Unless someone protests.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -20519,6 +20519,8 @@ with pkgs;
 
   protobuf = protobuf3_19;
 
+  protobuf3_21 = callPackage ../development/libraries/protobuf/3.21.nix { };
+  protobuf3_20 = callPackage ../development/libraries/protobuf/3.20.nix { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:20523]: Do we actually need both or is 3.21 enough?

File: pkgs/development/python-modules/protobuf/default.nix [hunk 2]
```diff
@@ -1,5 +1,6 @@
 { buildPackages
 , lib
+, buildPackagesProtobuf ? buildPackages.protobuf
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/protobuf/default.nix:3]:
```suggestion
, buildPackagesProtobuf
```
Please move this to top-level.

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 3]
```diff
@@ -0,0 +1,108 @@
+# The cmake version of this build is meant to enable both cmake and .pc being exported
+# this is important because grpc exports a .cmake file which also expects for protobuf
+# to have been exported through cmake as well.
+{ lib, stdenv
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/protobuf/generic-v3-cmake.nix:4]:
```suggestion
{ lib
, stdenv
```

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+, abseil-cpp
+, buildPackages
+, cmake
+, fetchFromGitHub
+, fetchpatch
+, gtest
+, zlib
+, version, sha256
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/protobuf/generic-v3-cmake.nix:12]:
```suggestion
, version
, sha256
```

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+, ...
+}:
+
+let
+  mkProtobufDerivation = buildProtobuf: stdenv: stdenv.mkDerivation {
+    pname = "protobuf";
+    inherit version;
+
+    # make sure you test also -A pythonPackages.protobuf
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/protobuf/generic-v3-cmake.nix:25]:
```suggestion
    # make sure you to also test nix-build -A pythonPackages.protobuf
```

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+      cp -r ${gtest.src}/googlemock third_party/gmock
+      cp -r ${gtest.src}/googletest third_party/
+      chmod -R a+w third_party/
+
+      ln -s ../googletest third_party/gmock/gtest
+      ln -s ../gmock third_party/googletest/googlemock
+      ln -s $(pwd)/third_party/googletest third_party/googletest/googletest
+    '' + lib.optionalString stdenv.isDarwin ''
+      substituteInPlace src/google/protobuf/testing/googletest.cc \
+      --replace 'tmpnam(b)' '"'$TMPDIR'/foo"'
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/protobuf/generic-v3-cmake.nix:44]:
```suggestion
        --replace 'tmpnam(b)' '"'$TMPDIR'/foo"'
```

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+      # https://github.com/protocolbuffers/protobuf/pull/10090
+      (fetchpatch {
+        url = "https://github.com/protocolbuffers/protobuf/commit/a7324f88e92bc16b57f3683403b6c993bf68070b.patch";
+        sha256 = "sha256-SmwaUjOjjZulg/wgNmR/F5b8rhYA2wkKAjHIOxjcQdQ=";
+      })
+    ];
+
+    nativeBuildInputs = [
+      cmake
+      buildPackages.which
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/protobuf/generic-v3-cmake.nix:58]:
```suggestion
      which
```
Why are we not doing this?

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    # a stub cmake/CMakeLists.txt still exists for compatibility with previous build assumptions
+    cmakeDir = "../cmake";
+    cmakeFlags = [
+      "-Dprotobuf_ABSL_PROVIDER=package"
+      ] ++ lib.optionals (!stdenv.targetPlatform.isStatic) [
+      "-Dprotobuf_BUILD_SHARED_LIBS=ON"
+    ];
+
+    enableParallelBuilding = true;
+
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/protobuf/generic-v3-cmake.nix:78]:
```suggestion
```
cmake already enabled this

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/top-level/all-packages.nix:20523]: I would prefer both, as you may need to introduce a specific version for some dev environments.

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix (hunk 5 above)
Comment (jonringer) [reply pkgs/development/libraries/protobuf/generic-v3-cmake.nix]: was removed

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix (hunk 7 above)
Comment (jonringer) [reply pkgs/development/libraries/protobuf/generic-v3-cmake.nix]: copied it from the other cmake builder, but it's no longer necessary

File: pkgs/development/python-modules/protobuf/default.nix (hunk 2 above)
Comment (jonringer) [reply pkgs/development/python-modules/protobuf/default.nix:3]: this was removed, instead using version as a way to find the correct package

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    meta = {
+      description = "Google's data interchange format";
+      longDescription = ''
+        Protocol Buffers are a way of encoding structured data in an efficient
+        yet extensible format. Google uses Protocol Buffers for almost all of
+        its internal RPC protocols and file formats.
+      '';
+      license = lib.licenses.bsd3;
+      platforms = lib.platforms.unix;
+      homepage = "https://developers.google.com/protocol-buffers/";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/protobuf/generic-v3-cmake.nix:103]: We should set meta.maintainers

File: pkgs/development/libraries/protobuf/generic-v3-cmake.nix [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+      "-Dprotobuf_ABSL_PROVIDER=package"
+      ] ++ lib.optionals (!stdenv.targetPlatform.isStatic) [
+      "-Dprotobuf_BUILD_SHARED_LIBS=ON"
+    ];
+
+    # unfortunately the shared libraries have yet to been patched by nix, thus tests will fail
+    doCheck = false;
+
+    passthru = {
+      inherit version;
```
[the remark this answers is not in the document]
Comment (jonringer) [reply pkgs/development/libraries/protobuf/generic-v3-cmake.nix]: removed

**Comment by jonringer** [comment]
Was also able to verify that the cross builds still work by doing `nix-build -A pkgsCross.aarch64-multiplatform.protobuf3_21`

**Comment by jonringer** [comment]
accidentally pushed from the wrong computer

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by jonringer** [comment]
Going to merge this as it has almost been 3 months. Additional fixes can be handled in later PRs

**Pull request merged**